File: python_implementation/FordFulkerson.py
import collections

#algorithm from https://en.wikipedia.org/wiki/Ford%E2%80%93Fulkerson_algorithm

class GraphAlgorithm:

    def __init__(self,edgeCount):
        self.edgeCount = edgeCount
        self.residualGraph = []
        for i in range(0,self.edgeCount):
            self.residualGraph.append( [0]*edgeCount)

        self.capacityGraph = []
        for i in range(0,self.edgeCount):
            self.capacityGraph.append( [0]*edgeCount)

    # assuming directed edge from node1 -> node2 with capacity _capacity_
    # and flow _flow_
    def addEdge(self,node1,node2,capacity,flow = 0):
        self.residualGraph[node1][node2] = capacity #- flow
        self.residualGraph[node2][node1] = flow

        # in directed graph, helps keep track of which direction is the capacity
        self.capacityGraph[node1][node2] = capacity

    # return boolean based on if it is possible to reach
    # node _finish_ if starting at node _finish_. Algorithm used 
    # is breadth first search (bfs). If boolean returned, then 
    # second value returned is a parent array
    def breadthFirstSearch(self,adjacencyMatrix,start,finish):
        visited = [False] * self.edgeCount
        path = [-1]*self.edgeCount

        queue = collections.deque()

        queue.append(start)
        visited[start] = True

        while queue:
            currentNode = queue.popleft()

            #find all neighbours of currentNode and add them to queue
            for potentialNeighbour in range(0,self.edgeCount): 
                if adjacencyMatrix[currentNode][potentialNeighbour] > 0 and visited[potentialNeighbour] == False:
                    visited[potentialNeighbour] = True 
                    queue.append(potentialNeighbour)
                    path[potentialNeighbour] = currentNode 

        return visited[finish],path

    def EdmondsKarp(self,source,sink):    
        adjacencyMatrixCopy = []
        for i in range(0,self.edgeCount):
            adjacencyMatrixCopy.append([0]*self.edgeCount)
        for i in range(0,self.edgeCount):
            for j in range(0,self.edgeCount):
                adjacencyMatrixCopy[i][j] = self.residualGraph[i][j]
        flow = 0

        while True:

            canReachSink, parentArray = self.breadthFirstSearch(adjacencyMatrixCopy,source,sink)
            if canReachSink == False:
                break

            path_flow = float("Inf")
            currentNode = sink
            while currentNode != source:
                path_flow = min( path_flow, adjacencyMatrixCopy[parentArray[currentNode]][currentNode] )
                currentNode = parentArray[currentNode]

            flow = flow + path_flow

            currentNode = sink
            while currentNode != source:
                adjacencyMatrixCopy[parentArray[currentNode]][currentNode] = adjacencyMatrixCopy[parentArray[currentNode]][currentNode] - path_flow
                adjacencyMatrixCopy[currentNode][parentArray[currentNode]] = adjacencyMatrixCopy[currentNode][parentArray[currentNode]] + path_flow
                currentNode = parentArray[currentNode]

        # compute the flow graph:
        flowGraph = []
        for i in range(0,self.edgeCount):
            flowGraph.append([0]*self.edgeCount)
        
        for i in range(0,self.edgeCount):
            for j in range(0,self.edgeCount):
                if self.capacityGraph[i][j] > 0: 
                    flowGraph[i][j] = adjacencyMatrixCopy[j][i] 

        return flow,flowGraph

File: java_implementation/testing/fordfulkerson.java
import java.util.Queue; 
import java.util.LinkedList;

public class fordfulkerson{

	int noParent = -1;
	int flow = 0;
	int edgeCount = 0;
	int [][] residualGraph = null;
	int [][] capacityGraph = null;

	public fordfulkerson(int edgeCount){
		this.edgeCount = edgeCount;
		this.residualGraph = new int[edgeCount][];
		this.capacityGraph = new int[edgeCount][];
		for( int i = 0; i < edgeCount; i++ ){
			int [] arr = new int[edgeCount];
			int [] arr2 = new int[edgeCount];
			for( int j = 0; j < edgeCount; j++ ){
				arr[j] = 0;
				arr2[j] = 0;
			}
			this.residualGraph[i] = arr;
			this.capacityGraph[i] = arr2;
		}

	}

	public int getFlow(){
		return this.flow;
	}

	public void addEdge(int node1, int node2, int capacity, int flow){
		this.residualGraph[node1][node2] = capacity - flow;
		this.residualGraph[node2][node1] = flow;
		this.capacityGraph[node1][node2] = capacity;

	}

	public int [] breadthFirstSearch(int[][] residualGraph,int startNode){

		boolean[] visited = new boolean[this.edgeCount];
		int[] path = new int[this.edgeCount];
		for( int i = 0; i < this.edgeCount; i++ ){
			visited[i] = false;
			path[i] = this.noParent;
		}

		Queue<Integer> q = new LinkedList<Integer>();
		q.add(startNode);

		while(q.size() > 0){
			int currentNode = q.remove();
			for(int i = 0; i < this.edgeCount; i++){
				if( residualGraph[currentNode][i] > 0 && visited[i] == false){
					visited[i] = true;
					q.add(i);
					path[i] = currentNode;
				}
			}
		}
		return path;
	}

	public int[][] EdmondsKarp( int source, int sink ){
		
		int [][] residualGraphCopy = new int[edgeCount][];
		for( int i = 0; i < edgeCount; i++ ){
			int [] arr = new int[edgeCount];
			for( int j = 0; j < edgeCount; j++ ){
				arr[j] = this.residualGraph[i][j];
			}
			residualGraphCopy[i] = arr;
		}

		int flow = 0;
		
		while(true){
			int [] path = this.breadthFirstSearch( residualGraphCopy, source);
			boolean canReachSink = path[sink] != this.noParent;

			if(canReachSink == false){
				break;
			}
			
			int path_flow = Integer.MAX_VALUE;
			int currentNode = sink;

			while( currentNode != source){
				path_flow = Math.min(path_flow,residualGraphCopy[path[currentNode]][currentNode]);
				currentNode = path[currentNode];
			}

			flow = flow + path_flow;
			
			currentNode = sink;
			while( currentNode != source ){
				residualGraphCopy[path[currentNode]][currentNode] = residualGraphCopy[path[currentNode]][currentNode] - path_flow;
				residualGraphCopy[currentNode][path[currentNode]] = residualGraphCopy[currentNode][path[currentNode]] + path_flow;
				currentNode = path[currentNode];
			}
		}


		int [][] flowGraph = new int[edgeCount][];
		for( int i = 0; i < edgeCount; i++ ){
			int [] arr = new int[edgeCount];
			for( int j = 0; j < edgeCount; j++ ){
				arr[j] = 0;
			}
			flowGraph[i] = arr;
		}

		
		for(int i = 0; i < this.edgeCount; i++){
			for(int j = 0; j < this.edgeCount; j++){
				if( this.capacityGraph[i][j] > 0 ){
					flowGraph[i][j] = residualGraphCopy[j][i];
				}
			}
		}

		this.flow = flow;
		return flowGraph;

	}
}
